### PR TITLE
Caching mono run file, loading the cache file for duo runs

### PR DESCRIPTION
**Pull request by wongalvis14**
Correctness tested by matching metrics on small dev set.
Mono cache will include the rankings of all texts, so for subsequent duo runs any number of hits would work.

Ran in hydra, Python 3.8.8

### CODE REVIEW

File: pygaggle/model/evaluate.py [hunk 1]
```diff
@@ -213,6 +218,8 @@ def evaluate(self,
             mono_out = self.mono_reranker.rerank(example.query, example.documents)
             mono_texts.append(sorted(enumerate(mono_out), key=lambda x: x[1].score, reverse=True)[:self.mono_hits])
             scores.append(np.array([x.score for x in mono_out]))
+            if self.mono_cache_writer is not None:
```
Comment (rodrigonogueira4) [inline pygaggle/model/evaluate.py:221]: nit: the line below is equivalent and more concise.
```suggestion
            if self.mono_cache_writer:
```

**Review by rodrigonogueira4 — APPROVED** [review]
LGTM

**Comment by wongalvis14** [comment]
The part left WIP is "Loading the 'cache' run file"

Although the rankings in the run file are preserved when loaded, their scores are not recorded in the run file. When the metrics are calculated (called "accumulated" in the code), the document rankings are deduced from the sorted order of the scores. As a result, if the number of documents reranked in the duo stage is `n`, any document beyond the `n`th will have a score of 0 and the metrics beyond the `n`th documents will be inaccurate. 

I wonder if there is an alternate format of a run file that records the score, or if there is an existing method to inject the scores based on their pre-existing ranking in a list of documents.

**Comment by wongalvis14** [comment]
It was working until I merged the branches recently. I suspect that it was due to changing `rerank` to `rescore`. Right now, if we change `rescore` back to `rerank` in Mono and Duo, it'll work just like before. But if we change them to `rerank`, even if we sort them, they (a single mono + duo run, and a duo run with cached mono run results) will give different results.

File: pygaggle/model/evaluate.py [hunk 2]
```diff
@@ -12,7 +14,7 @@
 from pygaggle.data.retrieval import RetrievalExample
 from pygaggle.rerank.base import Reranker
 from pygaggle.qa.base import Reader
-from pygaggle.model.writer import Writer
+from pygaggle.model.writer import (Writer, MsMarcoWriter)
```
Comment (ronakice) [inline pygaggle/model/evaluate.py:17]: no need the brackets

File: pygaggle/data/msmarco.py [hunk 3]
```diff
@@ -79,11 +80,14 @@ def load_queries(cls,
...
-                    is_duo: bool = False) -> 'MsMarcoDataset':
+                    is_duo: bool = False,
+                    run_path: Path = '.') -> 'MsMarcoDataset':
         run_mono = "mono." if is_duo else ""
         query_path = os.path.join(folder, f"queries.{split}.small.tsv")
         qrels_path = os.path.join(folder, f"qrels.{split}.small.tsv")
-        run_path = os.path.join(folder, f"run.{run_mono}{split}.small.tsv")
+        if not os.path.isfile(run_path):
+            run_path = os.path.join(folder, f"run.{run_mono}{split}.small.tsv")
+        print(run_path)
```
Comment (ronakice) [inline pygaggle/data/msmarco.py:90]: Can this print statement add a little more context, maybe two depending on if the file exists? Also it should be using logging I think?

File: pygaggle/run/evaluate_passage_ranker.py [hunk 4]
```diff
@@ -156,6 +157,14 @@ def main():
                      default=50,
                      help='Top k candidates from mono for duo reranking'),
                  opt('--output-file', type=Path, default='.'),
+                 opt('--mono-cache-dir',
+                     type=Path,
+                     default='.',
+                     help='Path to folder where the mono cache run files will be stored'),
```
[the remark this answers is not in the document]
Comment (ronakice) [reply pygaggle/run/evaluate_passage_ranker.py:163]: Also can we not fold this into mono-cache-run-path? If the file doesn't exist we create it and store results there, in which case this directory is not needed?

**Review by ronakice — APPROVED** [review]
LGTM! Merging.

**Pull request merged**